File: src/main/kotlin/com/wynnlab/spells/archer/ArrowStorm.kt
package com.wynnlab.spells.archer

import com.wynnlab.random
import com.wynnlab.spells.Spell
import com.wynnlab.spells.SpellData
import com.wynnlab.util.times
import org.bukkit.Material
import org.bukkit.Particle
import org.bukkit.Sound
import org.bukkit.entity.Arrow
import org.bukkit.entity.Player
import org.bukkit.entity.Projectile
import org.bukkit.entity.Snowball
import org.bukkit.inventory.ItemStack

class ArrowStorm(player: Player) : Spell(player, 21, SpellData.ARROW_STORM) {
    private val arrows = mutableListOf<Projectile>()

    override fun tick() {
        if (tick == 0)
            player.spawnParticle(if (clone) Particle.VILLAGER_HAPPY else Particle.CRIT, player.location, 5)

        player.playSound(player.location, Sound.ENTITY_ARROW_SHOOT, 1f, 1f)
        if (clone) player.playSound(player.location, Sound.BLOCK_GLASS_BREAK, .05f, .6f)

        for (arrow in arrows) {
            player.spawnParticle(Particle.FLAME, arrow.location, 1, .0, .0, .0, .01)
            player.spawnParticle(if (clone) Particle.VILLAGER_HAPPY else Particle.CRIT, arrow.location, 2, .0, .0, .0, .0)
        }

        val arrowM = player.launchProjectile(if (clone) Snowball::class.java else Arrow::class.java, player.eyeLocation.direction * 3.0)
        val arrowL = player.launchProjectile(if (clone) Snowball::class.java else Arrow::class.java, (player.eyeLocation.direction * 3.0).rotateAroundY(.4))
        val arrowR = player.launchProjectile(if (clone) Snowball::class.java else Arrow::class.java, (player.eyeLocation.direction * 3.0).rotateAroundY(-.4))

        for (arrow in arrayOf(arrowM, arrowL, arrowR)) {
            arrow.shooter = player
            arrow.addScoreboardTag("storm_arrow")
            arrows.add(arrow)
            if (clone) {
                (arrow as Snowball).item = ItemStack(Material.FLINT)
                arrow.setRotation(random.nextFloat(), 0f)
            }
        }
    }
}

File: src/main/resources/classes/archer/arrow_storm.py
from org.bukkit import Material, Particle, Sound
from org.bukkit.entity import Arrow, Player, Snowball
from org.bukkit.inventory import ItemStack

from com.wynnlab.spells import PySpell

class Spell(PySpell):
    def __init__(self):
        self.arrows = set()

    def tick(self):
        if self.t == 0:
            self.particle(self.player.getLocation().clone().add(0, 1, 0), Particle.VILLAGER_HAPPY if self.clone else Particle.CRIT, 5, .2, 2, .2, .5)

        self.sound(Sound.ENTITY_ARROW_SHOOT, 1, 1)
        if self.clone:
            self.sound(Sound.BLOCK_GLASS_BREAK, .05, .6)

        for arrow in self.arrows:
            self.particle(arrow.getLocation(), Particle.CRIT, 1, 0, 0, 0, .01)
            if self.clone:
                self.particle(arrow.getLocation(), Particle.CRIT_MAGIC, 1, 0, 0, 0, .01)
        
        vi = self.player.getEyeLocation().getDirection().multiply(3)
        shots = (
            self.player.launchProjectile(Snowball if self.clone else Arrow, vi),
            self.player.launchProjectile(Snowball if self.clone else Arrow, vi.rotateAroundY(.4)),
            self.player.launchProjectile(Snowball if self.clone else Arrow, vi.rotateAroundY(-.8))
        )
        
        for arrow in shots:
            arrow.setShooter(self.player)
            arrow.addScoreboardTag('storm_arrow')
            if self.clone:
                arrow.setItem(ItemStack(Material.FLINT))

            self.arrows.add(arrow)

def delete_arrow(event):
    hit = event.getHitEntity()
    if not hit is None and not isinstance(hit, Player):
        PySpell.damage(event.getEntity().getShooter(), hit, False, .1, .6, 0, .25, 0, .15, 0)

    event.getEntity().remove()

PySpell.registerProjectileHit('storm_arrow', delete_arrow)